src/components/common/WalletProvider/index.tsx

```diff
@@ -1,13 +1,49 @@
-import { createContext, type ReactElement, type ReactNode, useEffect, useState } from 'react'
+import { createContext, type ReactElement, type ReactNode, useEffect, useState, useMemo } from 'react'
 import useOnboard, { type ConnectedWallet, getConnectedWallet } from '@/hooks/wallets/useOnboard'
+import useAsync from '@/hooks/useAsync'
+import { getSafeInfo } from '@safe-global/safe-gateway-typescript-sdk'
+import { useWeb3ReadOnly } from '@/hooks/wallets/web3'
+import { useCurrentChain } from '@/hooks/useChains'
+import { useRouter } from 'next/router'
+import ExternalStore from '@/services/ExternalStore'
+import { type Eip1193Provider } from 'ethers'
+import { getNestedWallet } from '@/utils/nested-safe-wallet'
 
-export const WalletContext = createContext<ConnectedWallet | null>(null)
+const { setStore, useStore } = new ExternalStore<string | null>(null)
+
+export const setNestedSafeAddress = setStore
+export const useNestedSafeAddress = useStore
```

Comment:
I agree. I am generally not happy with how that works as of now.
The problem is that the selected signer should ideally only live within our transaction modal context. But now it lives globally and with your suggestion it still lives inside the `WalletProvider`. 

Because as of now we have to be careful to always reset it to `undefined` when the modal is closed as we otherwise still use the selected Safe as signer and some buttons turn disabled as we think your parent Signer is your connected signer.

I will give it a second thought.